File: core/src/main/java/com/emarsys/core/activity/CurrentActivityWatchdog.java
package com.emarsys.core.activity;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;
import android.util.Log;

import com.emarsys.core.provider.Property;
import com.emarsys.core.util.Assert;


public class CurrentActivityWatchdog implements Application.ActivityLifecycleCallbacks {

    private Property<Activity> currentActivityProvider;

    public CurrentActivityWatchdog(Property<Activity> currentActivityProvider) {
        Assert.notNull(currentActivityProvider, "CurrentActivityProvider must not be null!");
        this.currentActivityProvider = currentActivityProvider;
    }

    @Override
    public void onActivityCreated(Activity activity, Bundle savedInstanceState) {

    }

    @Override
    public void onActivityStarted(Activity activity) {

    }

    @Override
    public void onActivityResumed(Activity activity) {
        currentActivityProvider.set(activity);
    }

    @Override
    public void onActivityPaused(Activity activity) {
        if (currentActivityProvider.get() == activity) {
            currentActivityProvider.set(null);
        }
    }

    @Override
    public void onActivityStopped(Activity activity) {

    }

    @Override
    public void onActivitySaveInstanceState(Activity activity, Bundle outState) {

    }

    @Override
    public void onActivityDestroyed(Activity activity) {

    }
}

File: core/src/main/java/com/emarsys/core/activity/CurrentActivityWatchdog.kt
package com.emarsys.core.activity

import android.app.Activity
import android.app.Application.ActivityLifecycleCallbacks
import android.os.Bundle
import com.emarsys.core.Mockable
import com.emarsys.core.provider.Property

@Mockable
class CurrentActivityWatchdog(private val currentActivityProvider: Property<Activity?>) : ActivityLifecycleCallbacks {
    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {}
    override fun onActivityStarted(activity: Activity) {}
    override fun onActivityResumed(activity: Activity) {
        currentActivityProvider.set(activity)
    }

    override fun onActivityPaused(activity: Activity) {
        if (currentActivityProvider.get() === activity) {
            currentActivityProvider.set(null)
        }
    }

    override fun onActivityStopped(activity: Activity) {}
    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}
    override fun onActivityDestroyed(activity: Activity) {}
}